### ISSUE TITLE
Support for NPC

### ISSUE THREAD

**Issue by AbyssMorgan** [samp-incognito/samp-streamer-plugin#199]
Hi, current dynamic areas/cp/racecp not detecting NPC players.

**Comment by rt-2** [comment]
temporary solution, do your own different test, check ``IsPlayeNPC``, if yes check position and ``IsPointInDynamicArea`` to see if in the area with the position of hat NPC.

**Comment by AbyssMorgan** [comment]
I don't need check if npc is in dynamic area.

**Comment by rt-2** [comment]
Then why this issue? I'm not sure to understand...

**Comment by AbyssMorgan** [comment]
I need detect on npc enter/leave area

**Comment by rt-2** [comment]
Yes you will do exactly like this plugin does, you are gonna do a a timer, and ever x time you will check if a NPC entered a area, and if they did,,, then execute wanted code.

**Comment by samp-incognito** [comment]
I don't think I ever really tested compatibility with NPCs. Did it work in v2.8.2?

**Comment by AbyssMorgan** [comment]
In v2.9.2 not working and i thing in older version also not working.

**Comment by IstuntmanI** [comment]
Looks like ```sampgdk::IsPlayerNPC``` is used only 3 times in this plugin, so the plugin won't stream objects/mapIcons/3DTextLabels for them. I never really worked with NPCs, but I can't see why those callbacks aren't called for you (I didn't test myself). Are you sure that you don't have some 
```C++
if( IsPlayerNPC( playerid ) ) return 1;
```
in your callbacks, before checking areaid ? Also, test ```IsPlayerIn(Any)DynamicArea```, it should return 1 for a player if he is in (any) area. (but, if what you said is correct, it should return 0) ```Streamer::processAreas``` does Enter/Leave callbacks calling and setting if the player is in area, could you add some debug prints to it and check where it stops ? If the enter callback wasn't called, the area isn't in the ```player.internalAreas```, so ```IsPlayerIn(Any)DynamicArea``` should indeed return 0.

About (Race)CPs, is SA-MP actually calling OnPlayerEnter(Race)Checkpoint (you could add a debug print to these callbacks in ```callbacks.cpp``` too) ? I don't know if NPCs can actually enter in (race) checkpoints.

If you can't come with any more detailed response to help fixing this bug, I'll try to set up a NPC and test these things.

**Comment by Patrik356b** [comment]
The only items that should stream for npc's are checkpoints, race checkpoints, pickups and gangzones/areas as the rest is just unnecessary.

**Comment by IstuntmanI** [comment]
@Patrik356b: This is what this plugin does already, with one exception:

@samp-incognito: You should change the line [461/streamer.cpp](https://github.com/samp-incognito/samp-streamer-plugin/blob/b5b5257b12c0669b78e8c72528a39fcbf01aea05/src/streamer.cpp#L461) to
```C++
if (!core->getData()->actors.empty() && player.enabledItems[STREAMER_TYPE_ACTOR] && !sampgdk::IsPlayerNPC(player.playerID))
```
there's no need to check actors if a NPC is near them.

Also, a little optimisation I saw: you should move the line [764/streamer.cpp](https://github.com/samp-incognito/samp-streamer-plugin/blob/b5b5257b12c0669b78e8c72528a39fcbf01aea05/src/streamer.cpp#L764)
right above
```C++
for (std::vector<SharedCell>::const_iterator c = cells.begin(); c != cells.end(); ++c)
```
so it won't be called more than once.

I could give a pull request for this, but I guess that it would be too much just for these little changes.

----------

I just tested this issue with debug prints in lots of places. I tested it with the FCNPC plugin.

Something weird happens: NPCs aren't in the ```players``` container. ```OnPlayerConnect``` isn't called in the Streamer Plugin (via sampGDK), but the same debug line in my filterscript's ```OnPlayerConnect``` was printed out, so this means that NPCs are calling ```OnPlayerConnect``` only in filterscripts (and probably the gamemode), but not in sampGDK. Can somebody confirm this ? ( @Zeex / @ziggi ) As I said, I didn't work too much with NPCs before, so I don't know if this is a bug from FCNPC ([this line](https://github.com/ziggi/FCNPC/blob/1e969b37a6d1a350aae0d440dc07a2dd5e84f5ed/src/CFunctions.cpp#L97) adds "NPCs", so this plugin isn't using SA:MP's default NPC system I think), SA-MP or sampGDK.

So after that I added this in Streamer Plugin's ```startAutomaticUpdate``` function (I know, slow, but it doesn't matter in this case):
```C++
for(int i = 0; i < MAX_PLAYERS; i++)
{
	if( !sampgdk::IsPlayerConnected( i ) )
		continue;

	if( core->getData()->players.find( i ) == core->getData()->players.end( ) )
	{
		sampgdk::logprintf("Streamer: Player %d is connected but he isn't in the players container - adding him now | IsNPC: %d", i, sampgdk::IsPlayerNPC(i));
		Player player(i);
		core->getData()->players.insert(std::make_pair(i, player));
	}
}
```
and these lines were printed from the ```OnPlayerEnterDynamicArea``` callback:
```C++
Player 0 entered in areaid 1 // me

// NPCs
Player 30 entered in areaid 1
Player 31 entered in areaid 1
Player 32 entered in areaid 1
Player 33 entered in areaid 1
Player 34 entered in areaid 1
Player 35 entered in areaid 1
Player 36 entered in areaid 1
Player 37 entered in areaid 1
Player 38 entered in areaid 1
Player 39 entered in areaid 1
Player 40 entered in areaid 1
Player 41 entered in areaid 1
Player 42 entered in areaid 1
Player 43 entered in areaid 1
Player 44 entered in areaid 1
Player 45 entered in areaid 1
Player 46 entered in areaid 1
Player 47 entered in areaid 1
Player 48 entered in areaid 1
Player 49 entered in areaid 1
```
which is the expected behavior.

**Comment by samp-incognito** [comment]
I guess that makes sense. It might be a good idea for someone to open an issue there (https://github.com/Zeex/sampgdk) about this if sampgdk is at fault. I'm not sure why that would be the case, though.

**Comment by ziggi** [comment]
> this line adds "NPCs", so this plugin isn't using SA:MP's default NPC system I think

No, this function calls [this function](https://github.com/Sasuke78200/open-samp/blob/master/Open%20SAMP/RPC.cpp#L211) on the server. This means, what if you see `[npc:join] %s has joined the server (%u:%s)` in your logs, then OnPlayerConnect is called.

**Comment by IstuntmanI** [comment]
Well, yeah, I meant that it is the function that creates the NPC, as it have these lines:
https://github.com/ziggi/FCNPC/blob/1e969b37a6d1a350aae0d440dc07a2dd5e84f5ed/src/CFunctions.cpp#L124-L126

-------

Some guy told me that regular NPCs detect checkpoints and dynamic areas just fine, so they call ```OnPlayerConnect``` like they should, so it is probably a FCNPC problem, somehow.
